KNeighborsClassifier = 3

In [ ]:
import pandas as pd
from sklearn.cluster import KMeans
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report


In [ ]:
# Read the CSV file
df = pd.read_csv('/content/drive/MyDrive/Project 4 MN Home-loan Predictive Model (HMDA Data-source)/\
Clean_Data_For_Model/Cleaned_loan_data_for_model.csv')

df.head()

activity_year                   lei state_code       county_name  \
0           2023  254900FKD0CVFUC78457         MN      Anoka County   
1           2023  254900FKD0CVFUC78457         MN  St. Louis County   
2           2023  254900FKD0CVFUC78457         MN  St. Louis County   
3           2023  254900FKD0CVFUC78457         MN  St. Louis County   
4           2023  254900FKD0CVFUC78457         MN  St. Louis County   

  conforming_loan_limit             derived_dwelling_category  \
0            Conforming  Single Family (1-4 Units):Site-Built   
1            Conforming  Single Family (1-4 Units):Site-Built   
2            Conforming  Single Family (1-4 Units):Site-Built   
3            Conforming  Single Family (1-4 Units):Site-Built   
4            Conforming  Single Family (1-4 Units):Site-Built   

        derived_ethnicity derived_race derived_sex  action_taken  ...  \
0  Not Hispanic or Latino        Asian        Male             1  ...   
1  Not Hispanic or Latino        White      Female             1  ...   
2  Not Hispanic or Latino        White        Male             1  ...   
3  Not Hispanic or Latino        White        Male             1  ...   
4  Not Hispanic or Latino        White        Male             1  ...   

               lien_status reverse_mortgage open-end_line_of_credit  \
0  Secured by a first lien           Exempt                  Exempt   
1  Secured by a first lien           Exempt                  Exempt   
2  Secured by a first lien           Exempt                  Exempt   
3  Secured by a first lien           Exempt                  Exempt   
4  Secured by a first lien           Exempt                  Exempt   

  loan_amount              hoepa_status income applicant_race-1  \
0    185000.0  Not a high-cost mortgage   52.0            Asian   
1     55000.0  Not a high-cost mortgage   58.0            White   
2    715000.0  Not a high-cost mortgage  337.0            White   
3    105000.0  Not a high-cost mortgage  169.0            White   
4     55000.0            Not applicable   93.0            White   

   applicant_sex denial_reason-1  tract_minority_population_percent  
0           Male             NaN                              10.74  
1         Female             NaN                               6.68  
2           Male             NaN                               5.94  
3           Male             NaN                              10.24  
4           Male             NaN                               6.11  

[5 rows x 24 columns]

In [ ]:
# Seperate the features, X,  from the target variable, y
y = df['action_taken']
X = df.drop(columns='action_taken')

In [ ]:
# Preview the features data
X.head()

activity_year                   lei state_code       county_name  \
0           2023  254900FKD0CVFUC78457         MN      Anoka County   
1           2023  254900FKD0CVFUC78457         MN  St. Louis County   
2           2023  254900FKD0CVFUC78457         MN  St. Louis County   
3           2023  254900FKD0CVFUC78457         MN  St. Louis County   
4           2023  254900FKD0CVFUC78457         MN  St. Louis County   

  conforming_loan_limit             derived_dwelling_category  \
0            Conforming  Single Family (1-4 Units):Site-Built   
1            Conforming  Single Family (1-4 Units):Site-Built   
2            Conforming  Single Family (1-4 Units):Site-Built   
3            Conforming  Single Family (1-4 Units):Site-Built   
4            Conforming  Single Family (1-4 Units):Site-Built   

        derived_ethnicity derived_race derived_sex  purchaser_type  ...  \
0  Not Hispanic or Latino        Asian        Male  Not applicable  ...   
1  Not Hispanic or Latino        White      Female  Not applicable  ...   
2  Not Hispanic or Latino        White        Male  Not applicable  ...   
3  Not Hispanic or Latino        White        Male  Not applicable  ...   
4  Not Hispanic or Latino        White        Male  Not applicable  ...   

               lien_status reverse_mortgage open-end_line_of_credit  \
0  Secured by a first lien           Exempt                  Exempt   
1  Secured by a first lien           Exempt                  Exempt   
2  Secured by a first lien           Exempt                  Exempt   
3  Secured by a first lien           Exempt                  Exempt   
4  Secured by a first lien           Exempt                  Exempt   

  loan_amount              hoepa_status income  applicant_race-1  \
0    185000.0  Not a high-cost mortgage   52.0             Asian   
1     55000.0  Not a high-cost mortgage   58.0             White   
2    715000.0  Not a high-cost mortgage  337.0             White   
3    105000.0  Not a high-cost mortgage  169.0             White   
4     55000.0            Not applicable   93.0             White   

  applicant_sex  denial_reason-1 tract_minority_population_percent  
0          Male              NaN                             10.74  
1        Female              NaN                              6.68  
2          Male              NaN                              5.94  
3          Male              NaN                             10.24  
4          Male              NaN                              6.11  

[5 rows x 23 columns]

In [ ]:
# Preview the first five entries for the target variable
y[:5]

0    1
1    1
2    1
3    1
4    1
Name: action_taken, dtype: int64

In [ ]:
# Encode the categorical variables using get_dummies
X = pd.get_dummies(X)

In [ ]:
# Review the features data
X.head()

activity_year  loan_amount  income  tract_minority_population_percent  \
0           2023     185000.0    52.0                              10.74   
1           2023      55000.0    58.0                               6.68   
2           2023     715000.0   337.0                               5.94   
3           2023     105000.0   169.0                              10.24   
4           2023      55000.0    93.0                               6.11   

   lei_1IE8VN30JCEQV1H4R804  lei_1TMVIO1SD0RLIPEIW803  \
0                     False                     False   
1                     False                     False   
2                     False                     False   
3                     False                     False   
4                     False                     False   

   lei_2138004QPZ7FMGPTU106  lei_213800M6RMI3YBX62W88  \
0                     False                     False   
1                     False                     False   
2                     False                     False   
3                     False                     False   
4                     False                     False   

   lei_213800QUAI2VH5YM6310  lei_213800THW6KSUL191M34  ...  \
0                     False                     False  ...   
1                     False                     False  ...   
2                     False                     False  ...   
3                     False                     False  ...   
4                     False                     False  ...   

   applicant_sex_Information not provided by applicant in mail, internet, or telephone application  \
0                                              False                                                 
1                                              False                                                 
2                                              False                                                 
3                                              False                                                 
4                                              False                                                 

   applicant_sex_Male  applicant_sex_Not applicable  \
0                True                         False   
1               False                         False   
2                True                         False   
3                True                         False   
4                True                         False   

   denial_reason-1_Collateral  denial_reason-1_Credit history  \
0                       False                           False   
1                       False                           False   
2                       False                           False   
3                       False                           False   
4                       False                           False   

   denial_reason-1_Debt-to-income ratio  denial_reason-1_Employment history  \
0                                 False                               False   
1                                 False                               False   
2                                 False                               False   
3                                 False                               False   
4                                 False                               False   

   denial_reason-1_Insufficient cash (downpayment, closing costs)  \
0                                              False                
1                                              False                
2                                              False                
3                                              False                
4                                              False                

   denial_reason-1_Not applicable  denial_reason-1_Other  
0                           False                  False  
1                           False                  False  
2                           False                  False  
3                         

In [ ]:
# Split the dataset using train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1)

In [ ]:
# Instantiate a StandardScaler instance
scaler = StandardScaler()

# Fit the training data to the standard scaler
X_scaler = scaler.fit(X_train)

# Transform the training data using the scaler
X_train_scaled = X_scaler.transform(X_train)

# Transform the testing data using the scaler
X_test_scaled = X_scaler.transform(X_test)

In [ ]:
# Import the KNeighborsClassifier module from sklearn
from sklearn.neighbors import KNeighborsClassifier

# Instantiate the KNeighborsClassifier model with n_neighbors = 3
knn = KNeighborsClassifier(n_neighbors=3)

In [ ]:
# Train the model using the training data
knn.fit(X_train_scaled, y_train)

KNeighborsClassifier(n_neighbors=3)

In [ ]:
# Create predictions using the testing data
y_pred = knn.predict(X_test_scaled)

In [ ]:
# Print the classification report comparing the testing data to the model predictions
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.70      0.49      0.58       128
           1       1.00      1.00      1.00     19439

    accuracy                           1.00     19567
   macro avg       0.85      0.75      0.79     19567
weighted avg       0.99      1.00      0.99     19567



Key Observations:
- Strong performance on the majority class (approvals)
- Room for improvement in identifying denials
- Significant class imbalance (128 denials vs 19,439 approvals)
- Model may benefit from techniques to address class imbalance

KNeighborsClassifier = 5

In [ ]:
# Split the dataset using train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1)

In [ ]:
# Instantiate a StandardScaler instance
scaler = StandardScaler()

# Fit the training data to the standard scaler
X_scaler = scaler.fit(X_train)

# Transform the training data using the scaler
X_train_scaled = X_scaler.transform(X_train)

# Transform the testing data using the scaler
X_test_scaled = X_scaler.transform(X_test)

In [ ]:
# Import the ` module from sklearn
from sklearn.neighbors import KNeighborsClassifier

# Instantiate the KNeighborsClassifier model with n_neighbors = 5
knn = KNeighborsClassifier(n_neighbors=5)

In [ ]:
# Train the model using the training data
knn.fit(X_train_scaled, y_train)

KNeighborsClassifier()

In [ ]:
# Create predictions using the testing data
y_pred = knn.predict(X_test_scaled)

In [ ]:
# Print the classification report comparing the testing data to the model predictions
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.72      0.39      0.51       128
           1       1.00      1.00      1.00     19439

    accuracy                           1.00     19567
   macro avg       0.86      0.69      0.75     19567
weighted avg       0.99      1.00      0.99     19567



KNeighborsClassifier = 7

In [ ]:
# Split the dataset using train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1)

In [ ]:
# Instantiate a StandardScaler instance
scaler = StandardScaler()

# Fit the training data to the standard scaler
X_scaler = scaler.fit(X_train)

# Transform the training data using the scaler
X_train_scaled = X_scaler.transform(X_train)

# Transform the testing data using the scaler
X_test_scaled = X_scaler.transform(X_test)

In [ ]:
# Import the KNeighborsClassifier module from sklearn
from sklearn.neighbors import KNeighborsClassifier

# Instantiate the KNeighborsClassifier model with n_neighbors = 7
knn = KNeighborsClassifier(n_neighbors=7)

In [ ]:
# Train the model using the training data
knn.fit(X_train_scaled, y_train)

KNeighborsClassifier(n_neighbors=7)

In [ ]:
# Create predictions using the testing data
y_pred = knn.predict(X_test_scaled)

In [ ]:
# Print the classification report comparing the testing data to the model predictions
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.77      0.34      0.48       128
           1       1.00      1.00      1.00     19439

    accuracy                           1.00     19567
   macro avg       0.88      0.67      0.74     19567
weighted avg       0.99      1.00      0.99     19567



KNeighborsClassifier = 9

In [ ]:
# Split the dataset using train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1)

In [ ]:
# Instantiate a StandardScaler instance
scaler = StandardScaler()

# Fit the training data to the standard scaler
X_scaler = scaler.fit(X_train)

# Transform the training data using the scaler
X_train_scaled = X_scaler.transform(X_train)

# Transform the testing data using the scaler
X_test_scaled = X_scaler.transform(X_test)

In [ ]:
# Import the KNeighborsClassifier module from sklearn
from sklearn.neighbors import KNeighborsClassifier

# Instantiate the KNeighborsClassifier model with n_neighbors = 3
knn = KNeighborsClassifier(n_neighbors=9)

In [ ]:
# Train the model using the training data
knn.fit(X_train_scaled, y_train)

KNeighborsClassifier(n_neighbors=9)

In [ ]:
# Create predictions using the testing data
y_pred = knn.predict(X_test_scaled)

In [ ]:
# Print the classification report comparing the testing data to the model predictions
print(classification_report(y_test, y_pred))

7 is the ideal number for the KNeighborsClassifier for this dataset